diff --git a/findlib.conf.in b/findlib.conf.in
index 261d2c8..461bafc 100644
--- a/findlib.conf.in
+++ b/findlib.conf.in
@@ -1,2 +1,3 @@
 destdir="@SITELIB@"
 path="@SITELIB@"
+ldconf="ignore"
\ No newline at end of file
diff --git a/src/findlib/Makefile b/src/findlib/Makefile
index 4fd3f81..5b9a81e 100644
--- a/src/findlib/Makefile
+++ b/src/findlib/Makefile
@@ -123,7 +123,7 @@ clean:
 install: all
 	mkdir -p "$(prefix)$(OCAML_SITELIB)/$(NAME)"
 	mkdir -p "$(prefix)$(OCAMLFIND_BIN)"
-	test $(INSTALL_TOPFIND) -eq 0 || cp topfind "$(prefix)$(OCAML_CORE_STDLIB)"
+	test $(INSTALL_TOPFIND) -eq 0 || cp topfind "$(prefix)$(OCAML_SITELIB)"
 	files=`$(SH) $(TOP)/tools/collect_files $(TOP)/Makefile.config \
 	findlib.cmi findlib.mli findlib.cma findlib.cmxa findlib$(LIB_SUFFIX) findlib.cmxs \
 	findlib_config.cmi findlib_config.ml topfind.cmi topfind.mli \
